aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/data/EncryptedKeyValueRepository.kt

```diff
@@ -22,36 +22,81 @@ import androidx.security.crypto.EncryptedSharedPreferences
     private val context: Context,
     private val sharedPreferencesName: String,
 ) : KeyValueRepository {
 
+    @VisibleForTesting
+    internal val sharedPreferences: SharedPreferences by lazy {
+        EncryptedSharedPreferences.create(
+            "$sharedPreferencesName.${getInstallationIdentifier(context, sharedPreferencesName)}",
+            MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
+            context,
+            AES256_SIV,
+            AES256_GCM
+        )
+    }
+
+    @VisibleForTesting
+    internal val editor: SharedPreferences.Editor by lazy {
+        sharedPreferences.edit()
+    }
+
     override fun put(dataKey: String, value: String?) {
-        with(getSharedPreferences().edit()) {
+        with(editor) {
             putString(dataKey, value)
             apply()
         }
     }
 
-    override fun get(dataKey: String): String? = getSharedPreferences().getString(dataKey, null)
+    override fun get(dataKey: String): String? = sharedPreferences.getString(dataKey, null)
 
     override fun remove(dataKey: String) {
-        with(getSharedPreferences().edit()) {
+        with(editor) {
             remove(dataKey)
             apply()
         }
     }
 
-    @VisibleForTesting
-    fun getSharedPreferences(): SharedPreferences {
-        return EncryptedSharedPreferences.create(
-            sharedPreferencesName,
-            MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
-            context,
-            AES256_SIV,
-            AES256_GCM
-        )
+    /**
+     * EncryptedSharedPreferences may have been backed up by the application, but will be unreadable due to the
+     * KeyStore record being lost. To prevent an unreadable EncryptedSharedPreferences, we append a suffix to the name
+     * with a UUID created in the noBackupFilesDir
+     */
+    @Synchronized
+    private fun getInstallationIdentifier(context: Context, keyValueRepoID: String): String {
+        val identifierFile = File(context.noBackupFilesDir, "$keyValueRepoID.installationIdentifier")
+        val previousIdentifier = getExistingInstallationIdentifier(identifierFile)
+
+        return previousIdentifier ?: createInstallationIdentifier(identifierFile)
+    }
+
+    /**
+     * Gets the existing installation identifier (if exists)
+     */
+    private fun getExistingInstallationIdentifier(identifierFile: File): String? {
+        return if (identifierFile.exists()) {
+            val identifier = identifierFile.readText()
+            identifier.ifBlank { null }
+        } else {
+            null
+        }
+    }
+
+    /**
+     * Creates a new installation identifier for the install
+     */
+    private fun createInstallationIdentifier(identifierFile: File): String {
+        val newIdentifier = UUID.randomUUID().toString()
+        try {
+            identifierFile.writeText(newIdentifier)
+        } catch (e: Exception) {
+            // Failed to write identifier to file, session will be forced to be in memory
```

Comment:
1. interesting! no warning for empty catch?
2. >  session will be forced to be in memory
How would this work?

Reply:
@banji180 @sktimalsina You are both commenting on the same area. I should probably try to add logger in here and add a warning. This should never happen in practice as the noBackupDir app directory should alway be present and writeable. If there were some fatal error, there is really no graceful way to recover from this.

We could:
* Crash the application
* Use in memory storage (which obviously wouldn't retain credentials on a restart. 

I would want to avoid crashing the application at all costs.

Reply:
> Use in memory storage (which obviously wouldn't retain credentials on a restart.

Should we set identifier as empty string in case of fatal error so that app can retain credentials on a restart?

Reply:
I don't think so. This would mean that we would end up in the same situation as the original crash if an uninstall/reinstall occurred. Once again, this should never really happen in practice on a functioning Android device.